File: clients/windows/gui/src/Nyxveil.Client.Core/CatalogModels.cs
using System.Text.Json.Serialization;

namespace Nyxveil.Client.Core;

/// <summary>DTOs matching Control Plane / Frozen Core catalog JSON tags.</summary>
public sealed class LocationDto
{
    [JsonPropertyName("location_id")]
    public string LocationId { get; set; } = "";

    [JsonPropertyName("country")]
    public string Country { get; set; } = "";

    [JsonPropertyName("country_code")]
    public string CountryCode { get; set; } = "";

    [JsonPropertyName("city")]
    public string City { get; set; } = "";

    [JsonPropertyName("display_name")]
    public string DisplayName { get; set; } = "";

    [JsonPropertyName("enabled")]
    public bool Enabled { get; set; }
}

public sealed class EndpointDto
{
    [JsonPropertyName("host")]
    public string Host { get; set; } = "";

    [JsonPropertyName("port")]
    public int Port { get; set; }

    [JsonPropertyName("profiles")]
    public IReadOnlyList<string> Profiles { get; set; } = Array.Empty<string>();

    [JsonPropertyName("ip_family")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? IpFamily { get; set; }
}

public sealed class HealthInfoDto
{
    [JsonPropertyName("healthy")]
    public bool Healthy { get; set; }

    [JsonPropertyName("latency_ms")]
    public double LatencyMs { get; set; }

    [JsonPropertyName("session_count")]
    public int SessionCount { get; set; }

    [JsonPropertyName("cpu_percent")]
    public double CpuPercent { get; set; }

    [JsonPropertyName("memory_percent")]
    public double MemoryPercent { get; set; }
}

public sealed class NodeRegistryEntryDto
{
    [JsonPropertyName("node_id")]
    public string NodeId { get; set; } = "";

    [JsonPropertyName("location_id")]
    public string LocationId { get; set; } = "";

    [JsonPropertyName("country")]
    public string Country { get; set; } = "";

    [JsonPropertyName("city")]
    public string City { get; set; } = "";

    [JsonPropertyName("display_name")]
    public string DisplayName { get; set; } = "";

    [JsonPropertyName("status")]
    public string Status { get; set; } = "";

    [JsonPropertyName("enabled")]
    public bool Enabled { get; set; }

    [JsonPropertyName("test_only")]
    public bool TestOnly { get; set; }

    [JsonPropertyName("draining")]
    public bool Draining { get; set; }

    [JsonPropertyName("protocol_version")]
    public ushort ProtocolVersion { get; set; }

    [JsonPropertyName("server_version")]
    public string ServerVersion { get; set; } = "";

    [JsonPropertyName("endpoints")]
    public IReadOnlyList<EndpointDto> Endpoints { get; set; } = Array.Empty<EndpointDto>();

    [JsonPropertyName("server_name")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? ServerName { get; set; }

    [JsonPropertyName("spki_pin")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public byte[]? SpkiPin { get; set; }

    [JsonPropertyName("capacity")]
    public int Capacity { get; set; }

    [JsonPropertyName("current_sessions")]
    public int CurrentSessions { get; set; }

    [JsonPropertyName("health")]
    public HealthInfoDto Health { get; set; } = new();

    [JsonPropertyName("last_seen")]
    public DateTime LastSeen { get; set; }
}

public sealed class CatalogDto
{
    [JsonPropertyName("version")]
    public string Version { get; set; } = "";

    [JsonPropertyName("locations")]
    public IReadOnlyList<LocationDto> Locations { get; set; } = Array.Empty<LocationDto>();

    [JsonPropertyName("nodes")]
    public IReadOnlyList<NodeRegistryEntryDto> Nodes { get; set; } = Array.Empty<NodeRegistryEntryDto>();

    [JsonPropertyName("issued_at")]
    public DateTime IssuedAt { get; set; }

    [JsonPropertyName("expires_at")]
    public DateTime ExpiresAt { get; set; }
}

public sealed class SignedCatalogDto
{
    [JsonPropertyName("catalog")]
    public CatalogDto Catalog { get; set; } = new();

    [JsonPropertyName("key_id")]
    public string KeyId { get; set; } = "";

    [JsonPropertyName("signature")]
    public byte[] Signature { get; set; } = Array.Empty<byte>();
}

File: clients/android/app/src/main/java/ru/nyxveil/android/catalog/CatalogModels.kt
package ru.nyxveil.android.catalog

data class LocationDto(
    val locationId: String = "",
    val country: String = "",
    val countryCode: String = "",
    val city: String = "",
    val displayName: String = "",
    val enabled: Boolean = false,
)

data class EndpointDto(
    val host: String = "",
    val port: Int = 0,
    val profiles: List<String> = emptyList(),
    val ipFamily: String? = null,
)

data class HealthInfoDto(
    val healthy: Boolean = false,
    val latencyMs: Double = 0.0,
    val sessionCount: Int = 0,
    val cpuPercent: Double = 0.0,
    val memoryPercent: Double = 0.0,
)

data class NodeRegistryEntryDto(
    val nodeId: String = "",
    val locationId: String = "",
    val country: String = "",
    val city: String = "",
    val displayName: String = "",
    val status: String = "",
    val enabled: Boolean = false,
    val testOnly: Boolean = false,
    val draining: Boolean = false,
    val protocolVersion: Int = 0,
    val serverVersion: String = "",
    val endpoints: List<EndpointDto> = emptyList(),
    val serverName: String? = null,
    val spkiPin: ByteArray? = null,
    val capacity: Int = 0,
    val currentSessions: Int = 0,
    val health: HealthInfoDto = HealthInfoDto(),
    val lastSeen: String = "",
)

data class CatalogDto(
    val version: String = "",
    val locations: List<LocationDto> = emptyList(),
    val nodes: List<NodeRegistryEntryDto> = emptyList(),
    val issuedAt: String = "",
    val expiresAt: String = "",
)

data class SignedCatalogDto(
    val catalog: CatalogDto = CatalogDto(),
    val keyId: String = "",
    val signature: ByteArray = ByteArray(0),
)

data class CatalogKeys(
    val issuer: String = "",
    val keys: Map<String, String> = emptyMap(),
    val updatedAt: Long = 0L,
)

data class LocationUi(
    val locationId: String,
    val displayName: String,
    val city: String,
    val country: String,
)

data class CatalogVerifyReport(
    val keyId: String,
    val issuedAtEpochMs: Long,
    val expiresAtEpochMs: Long,
    val nowEpochMs: Long,
    val remainingSeconds: Long,
    val signature: String,
    val temporalValidation: String,
)

class CatalogTemporalException(
    message: String,
    val report: CatalogVerifyReport,
) : Exception(message)

class CatalogSignatureException(message: String) : Exception(message)
